File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack=[]


        for ch in s:
            if ch=='(' or ch=='[' or ch=='{':
                stack.append(ch)
            else:
                if not stack:
                    return False  #if stack is empty

                temp=stack.pop()

                if(ch==')' and temp!='(') or (ch==']' and temp!='[') or (ch=='}' and temp!='{'):
                    return False

        return len(stack)==0
       
        
        

File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<int>st;
        int n=s.length();

        for(int i=0;i<n;i++){
            char ch=s[i];
            if(ch=='(' || ch=='[' || ch=='{'){
                st.push(ch);
            }
            else{
                if(st.size()==0) return false;

                char top=st.top();
                st.pop();

                if((ch==')'&&top!='(') || (ch==']'&&top!='[') || (ch=='}'&&top!='{')){
                    return false;
                }


            }
        }

        return st.empty();
    }
};